### PR TITLE
[community pipeline] Add RF-inversion Flux pipeline

### PR DESCRIPTION
**Pull request by linoytsaban**
This PR adds [rf-inversion](https://rf-inversion.github.io/) Flux pipeline to support real image editing & zero-shot stylization with flux. 
Thanks to community contributions 🙌🏻- this PR is adapted from the diffusers implementation https://github.com/raven38/rf_inversion by @raven38,  
as well as (https://github.com/DarkMnDragon/rf-inversion-diffuser) by @DarkMnDragon 

Note - `start_timestep` & `stop_timestep` were converted (from the original implementation) to be a float, representing the fraction of the total amount of steps

Example: 

1. get pipeline
```
!wget https://raw.githubusercontent.com/linoytsaban/diffusers/rf-inversion/examples/community/pipeline_flux_rf_inversion.py
# !wget https://raw.githubusercontent.com/huggingface/diffusers/main/examples/community/pipeline_flux_rf_inversion.py
```

2. load model
```
import torch
import diffusers
from pipeline_flux_rf_inversion import RFInversionFluxPipeline

pipe = RFInversionFluxPipeline.from_pretrained("black-forest-labs/FLUX.1-dev", 
                                               torch_dtype=torch.bfloat16)
pipe.to("cuda")
```

3.  load & invert image
```
from PIL import Image
image = Image.open("./example/cat.png").resize((1024, 1024))

inverted_latents, image_latents, latent_image_ids = pipe.invert(image=image, 
num_inversion_steps=28,
     gamma=0.5)
```

4. edit

```
pipe("portrait of a tiger", 
      inverted_latents=inverted_latents,
     image_latents=image_latents,
     latent_image_ids=latent_image_ids,
     start_timestep=0, 
     stop_timestep=.38,
     num_inference_steps=28,
     eta=0.9, 
    ).images[0]
```

Group 1-51

### CODE REVIEW

**Comment by DarkMnDragon** [comment]
Hi @linoytsaban, thank you for your contribution! Just a reminder: in the denoising loop, when editing the prompt after inversion, we need to interpolate between two velocities—flux velocity (with the edited prompt) and the target image velocity. You can refer to [this line](https://github.com/DarkMnDragon/rf-inversion-diffuser/blob/fc17128c4fa4212d8ca869cb3e34f3306eb5cf39/inversion_editing_cli.py#L252).

This is consistent with the original ComfyUI implementation, as seen [here](https://github.com/logtd/ComfyUI-Fluxtapoz/blob/e0b9195df74bd3659dea6ac5f733157ba31210c6/nodes/rectified_sampler_nodes.py#L75), and there are recommendations for the $\eta$ parameter [here](https://github.com/logtd/ComfyUI-Fluxtapoz/blob/main/README.md).

**Comment by linoytsaban** [comment]
@DarkMnDragon Thanks for the notes and for sharing your implementation! I didn't get to these changes yet in the initial commit, but I'm adding them in now :)

**Comment by sayakpaul** [comment]
@linoytsaban it seems like it's still WIP but LMK.

File: examples/community/pipeline_flux_rf_inversion.py [hunk 1]
```diff
@@ -0,0 +1,978 @@
...
+        # latents, latent_image_ids = self.prepare_latents(
+        #     batch_size * num_images_per_prompt,
+        #     num_channels_latents,
+        #     height,
+        #     width,
+        #     prompt_embeds.dtype,
+        #     device,
+        #     generator,
+        #     latents,
+        # )
```
Comment (apolinario) [inline examples/community/pipeline_flux_rf_inversion.py:765]: probably ok to delete?

**Comment by linoytsaban** [comment]
@sayakpaul will likely make some additional minor style/documentation changes but other than that I think it's ready for review 🙌🏻

**Comment by sayakpaul** [comment]
@linoytsaban sure, but I think it might be better to have it reviewed in a more complete stage.

**Comment by jorgemcgomes** [comment]
Hi @linoytsaban could you provide an updated example on how to run this pipeline?

I tried the code example you posted before and it doesn't work.

For example, in your example you pass stop_timestep as an int, but the pipeline expects this to be a fraction, so there's obviously something wrong or outdated.

**Comment by hj13-mtlab** [comment]
@linoytsaban  I wonder how can I do style reference using rf-inversion  with pipeline_flux_rf_inversion.py?

**Comment by linoytsaban** [comment]
@sayakpaul ready for review when you are 🤗

File: examples/community/pipeline_flux_rf_inversion.py [hunk 2]
```diff
@@ -0,0 +1,969 @@
+# Copyright 2024 Black Forest Labs and The HuggingFace Team. All rights reserved.
```
Comment (sayakpaul) [inline examples/community/pipeline_flux_rf_inversion.py:1]: Do we need to add licensing for the original RF inversion pipeline authors?

File: examples/community/pipeline_flux_rf_inversion.py [hunk 3]
```diff
@@ -0,0 +1,969 @@
...
+        >>> _,__,___ = pipe.invert(image=image, num_inversion_steps=28)
+
+        >>> edited_image = pipe(
+        ...     prompt="a portrait of a tiger",
+        ... ).images[0]
+        ```
+"""
+
+
+def calculate_shift(
```
Comment (sayakpaul) [inline examples/community/pipeline_flux_rf_inversion.py:75]: Add copied from here as well?

File: examples/community/pipeline_flux_rf_inversion.py [hunk 4]
```diff
@@ -0,0 +1,969 @@
...
+        self.vae_scale_factor = (
+            2 ** (len(self.vae.config.block_out_channels) - 1) if hasattr(self, "vae") and self.vae is not None else 8
+        )
+        self.image_processor = VaeImageProcessor(vae_scale_factor=self.vae_scale_factor)
+        self.tokenizer_max_length = (
+            self.tokenizer.model_max_length if hasattr(self, "tokenizer") and self.tokenizer is not None else 77
+        )
+        self.default_sample_size = 128
+
+    def _get_t5_prompt_embeds(
```
Comment (sayakpaul) [inline examples/community/pipeline_flux_rf_inversion.py:214]: "# Copied from ..." statements here and elsewhere.

File: examples/community/pipeline_flux_rf_inversion.py [hunk 5]
```diff
@@ -0,0 +1,969 @@
...
+            pooled_prompt_embeds=pooled_prompt_embeds,
+            device=device,
+            num_images_per_prompt=num_images_per_prompt,
+            max_sequence_length=max_sequence_length,
+            lora_scale=lora_scale,
+        )
+
+        # 4. Prepare latent variables
+        num_channels_latents = self.transformer.config.in_channels // 4
+        latents = self.inverted_latents
```
Comment (sayakpaul) [inline examples/community/pipeline_flux_rf_inversion.py:727]: Not seeing `inverted_latents` anywhere above. Could this be a mistake?

File: examples/community/pipeline_flux_rf_inversion.py [hunk 6]
```diff
@@ -0,0 +1,969 @@
...
+        latent_image_ids = self.latent_image_ids
+        image_latents = self.image_latents
```
Comment (sayakpaul) [inline examples/community/pipeline_flux_rf_inversion.py:729]: Didn't see these attributes above, either.

File: examples/community/pipeline_flux_rf_inversion.py [hunk 7]
```diff
@@ -0,0 +1,969 @@
...
+                t_i = 1 - t / 1000
+                dt = torch.tensor(1 / (len(timesteps) - 1), device=device)
+
+                if self.interrupt:
+                    continue
+
+                # broadcast to batch dimension in a way that's compatible with ONNX/Core ML
+                timestep = t.expand(latents.shape[0]).to(latents.dtype)
+
+                v_t = -self.transformer(
```
Comment (sayakpaul) [inline examples/community/pipeline_flux_rf_inversion.py:775]: Let's add an explanation on why we're negating this?

File: examples/community/pipeline_flux_rf_inversion.py [hunk 8]
```diff
@@ -0,0 +1,969 @@
...
+            guidance = guidance.expand(latents.shape[0])
+        else:
+            guidance = None
+
+        # 6. Denoising loop
+        with self.progress_bar(total=num_inference_steps) as progress_bar:
+            y_0 = image_latents.clone()
+            for i, t in enumerate(timesteps):
+                t_i = 1 - t / 1000
+                dt = torch.tensor(1 / (len(timesteps) - 1), device=device)
```
Comment (sayakpaul) [inline examples/community/pipeline_flux_rf_inversion.py:767]: Accompany with a one-line comment?

File: examples/community/pipeline_flux_rf_inversion.py [hunk 9]
```diff
@@ -0,0 +1,969 @@
...
+
+                v_t_cond = (y_0 - latents) / (1 - t_i)
+                eta_t = eta if start_timestep <= i < stop_timestep else 0.0
+                if start_timestep <= i < stop_timestep:
+                    # controlled vector field
+                    v_hat_t = v_t + eta * (v_t_cond - v_t)
+
+                else:
+                    v_hat_t = v_t
+                # SDE Eq: 17
```
Comment (sayakpaul) [inline examples/community/pipeline_flux_rf_inversion.py:795]: SDE equation from? Accompany with a better reference?

File: examples/community/pipeline_flux_rf_inversion.py [hunk 10]
```diff
@@ -0,0 +1,969 @@
...
+
+         >>> def download_image(url):
+        ...     response = requests.get(url)
+        ...     return PIL.Image.open(BytesIO(response.content)).convert("RGB")
+
+
+        >>> img_url = "https://www.aiml.informatik.tu-darmstadt.de/people/mbrack/tennis.jpg"
+        >>> image = download_image(img_url)
+
+        >>> _,__,___ = pipe.invert(image=image, num_inversion_steps=28)
```
Comment (sayakpaul) [inline examples/community/pipeline_flux_rf_inversion.py:66]: Where are the outputs used?

File: examples/community/pipeline_flux_rf_inversion.py [hunk 11]
```diff
@@ -0,0 +1,969 @@
...
+logger = logging.get_logger(__name__)  # pylint: disable=invalid-name
+
+EXAMPLE_DOC_STRING = """
+    Examples:
+        ```py
+        >>> import torch
+        >>> from diffusers import FluxPipeline
+
+        >>> pipe = FluxPipeline.from_pretrained("black-forest-labs/FLUX.1-schnell", torch_dtype=torch.bfloat16)
+        >>> pipe.to("cuda")
```
Comment (sayakpaul) [inline examples/community/pipeline_flux_rf_inversion.py:56]: Is this supposed to used with schenll? I had also expected to see the `custom_pipeline` argument being set.

File: examples/community/pipeline_flux_rf_inversion.py [hunk 12]
```diff
@@ -0,0 +1,969 @@
...
+                # controlled vector field
+                # Eq 8 dY_t = [u_t(Y_t) + γ(u_t(Y_t|y_1) - u_t(Y_t))]dt
+                u_hat_t_i = u_t_i + gamma * (u_t_i_cond - u_t_i)
+                Y_t = Y_t + u_hat_t_i * (sigmas[i] - sigmas[i + 1])
+                progress_bar.update()
+
+        self.inverted_latents = Y_t
+        self.latent_image_ids = latent_image_ids
+
+        return self.image_latents, Y_t, latent_image_ids
```
Comment (sayakpaul) [inline examples/community/pipeline_flux_rf_inversion.py:969]: Curious as to why do we need to assign them to `self`. Maybe we could clarify the order of steps one needs to take to use the inversion pipeline? That would clarify a lot of things.

File: examples/community/pipeline_flux_rf_inversion.py (hunk 5 above)
Comment (linoytsaban) [reply examples/community/pipeline_flux_rf_inversion.py:727]: they're used to set the latents in the editing denoising loop in [this](https://github.com/huggingface/diffusers/blob/144782a2f0ce1a3fa90a08f512b468e2a49dc5d5/examples/community/pipeline_flux_rf_inversion.py#L727) line

File: examples/community/pipeline_flux_rf_inversion.py (hunk 6 above)
Comment (linoytsaban) [reply examples/community/pipeline_flux_rf_inversion.py:729]: similarly: https://github.com/huggingface/diffusers/blob/144782a2f0ce1a3fa90a08f512b468e2a49dc5d5/examples/community/pipeline_flux_rf_inversion.py#L728

File: examples/community/pipeline_flux_rf_inversion.py (hunk 12 above)
Comment (linoytsaban) [reply examples/community/pipeline_flux_rf_inversion.py:969]: I was mostly following the design/flow we had for ledits++(as it's also an inversion based pipeline) where we keep the inverted latents assigned to self after calling `invert`. But I don't feel strongly about it

**Comment by sayakpaul** [comment]
@linoytsaban I see some unresolved comments regarding adding comments. Did I miss something?

File: examples/community/pipeline_flux_rf_inversion.py (hunk 12 above)
Comment (staturecrane) [reply examples/community/pipeline_flux_rf_inversion.py:979]: just popping in to say for purposes in production where we are storing the pipeline in memory in a worker, it's pretty undesirable to have any ephemeral state related to a single request stored on the pipeline instance. I think returning back the latents from the method then passing them back to the __call__ method is much preferable.

**Comment by linoytsaban** [comment]
@sayakpaul sorry for the false alarm! now all addressed 🤗

**Review by apolinario — APPROVED** [review]
Looks really good 👍

If anything, as RF-Inversion is an established technique with good traction (https://rf-inversion.github.io) I would say that maybe it could become soon (if not now) a proper pipeline. But as a community one imo we can merge it!

File: examples/community/pipeline_flux_rf_inversion.py [hunk 13]
```diff
@@ -0,0 +1,1046 @@
...
+        >>> image = download_image(img_url)
+
+        >>> _,__,___ = pipe.invert(image=image, num_inversion_steps=28, gamma=0.5)
+
+        >>> edited_image = pipe(
+        ...     prompt="a tomato",
+        ...     start_timestep=0,
+        ...     stop_timestep=.38,
+        ...     num_inference_steps=28,
+        ...     eta=0.9,
```
Comment (hlky) [inline examples/community/pipeline_flux_rf_inversion.py:79]:
```suggestion
        >>> inverted_latents, image_latents, latent_image_ids = pipe.invert(image=image, num_inversion_steps=28, gamma=0.5)

        >>> edited_image = pipe(
        ...     prompt="a tomato",
        ...     inverted_latents=inverted_latents,
        ...     image_latents=image_latents,
        ...     latent_image_ids=latent_image_ids,
        ...     start_timestep=0,
        ...     stop_timestep=.38,
        ...     num_inference_steps=28,
        ...     eta=0.9,
```

File: examples/community/pipeline_flux_rf_inversion.py [hunk 14]
```diff
@@ -0,0 +1,1046 @@
...
+        prompt_embeds = prompt_embeds.pooler_output
+        prompt_embeds = prompt_embeds.to(dtype=self.text_encoder.dtype, device=device)
+
+        # duplicate text embeddings for each generation per prompt, using mps friendly method
+        prompt_embeds = prompt_embeds.repeat(1, num_images_per_prompt)
+        prompt_embeds = prompt_embeds.view(batch_size * num_images_per_prompt, -1)
+
+        return prompt_embeds
+
+    def encode_prompt(
```
Comment (hlky) [inline examples/community/pipeline_flux_rf_inversion.py:319]:
```suggestion
    # Copied from diffusers.pipelines.flux.pipeline_flux.FluxPipeline.encode_prompt
    def encode_prompt(
```

File: examples/community/pipeline_flux_rf_inversion.py [hunk 15]
```diff
@@ -0,0 +1,1046 @@
...
+
+        _, seq_len, _ = prompt_embeds.shape
+
+        # duplicate text embeddings and attention mask for each generation per prompt, using mps friendly method
+        prompt_embeds = prompt_embeds.repeat(1, num_images_per_prompt, 1)
+        prompt_embeds = prompt_embeds.view(batch_size * num_images_per_prompt, seq_len, -1)
+
+        return prompt_embeds
+
+    def _get_clip_prompt_embeds(
```
Comment (hlky) [inline examples/community/pipeline_flux_rf_inversion.py:275]:
```suggestion
    # Copied from diffusers.pipelines.flux.pipeline_flux.FluxPipeline._get_clip_prompt_embeds
    def _get_clip_prompt_embeds(
```

File: examples/community/pipeline_flux_rf_inversion.py (hunk 12 above)
Comment (staturecrane) [reply examples/community/pipeline_flux_rf_inversion.py]: thank you for considering my comment and making the change! :smiley:

**Pull request merged**